bindings/rust/aws-lc-sys-template/build.rs

```diff
@@ -0,0 +1,194 @@
+}
+
+fn verify_fips_clang_version() -> (&'static str, &'static str) {
+    fn version(tool: &str) -> String {
+        let output = match Command::new(tool).arg("--version").output() {
+            Ok(o) => o,
+            Err(e) => {
+                eprintln!("warning: missing {}, trying other compilers: {}", tool, e);
+                // NOTE: hard-codes that the loop below checks the version
+                return String::new();
+            }
+        };
+        assert!(output.status.success());
+        let output = std::str::from_utf8(&output.stdout).expect("invalid utf8 output");
+        output.lines().next().expect("empty output").to_string()
+    }
+
+    const REQUIRED_CLANG_VERSION: &str = "7.0.1";
+    for (cc, cxx) in [
+        ("clang-7", "clang++-7"),
+        ("clang", "clang++"),
+        ("cc", "c++"),
+    ] {
+        let cc_version = version(cc);
+        if cc_version.contains(REQUIRED_CLANG_VERSION) {
+            assert!(
+                version(cxx).contains(REQUIRED_CLANG_VERSION),
+                "mismatched versions of cc and c++"
+            );
+            return (cc, cxx);
+        } else if cc == "cc" {
+            panic!(
+                "unsupported clang version \"{}\": FIPS requires clang {}",
+                cc_version, REQUIRED_CLANG_VERSION
+            );
+        } else if !cc_version.is_empty() {
+            eprintln!(
+                "warning: FIPS requires clang version {}, skipping incompatible version \"{}\"",
+                REQUIRED_CLANG_VERSION, cc_version
+            );
+        }
+    }
+    unreachable!()
+}
+
+fn test_cmake_command(executable: &str) -> bool {
+    if let Ok(output) = Command::new(executable).arg("--version").output() {
+        return output.status.success();
+    }
+    false
+}
+
+fn prepare_cmake_build(
+    build_fips: bool,
+    build_prefix: Option<&str>) -> cmake::Config {
+
+    if test_cmake_command("cmake3") {
+        env::set_var("CMAKE", "cmake3");
+    } else {
+        env::set_var("CMAKE", "cmake");
+    }
+
+    let mut cmake_cfg = get_cmake_config();
+
+    if build_fips {
+        let (clang, clangxx) = verify_fips_clang_version();
+        cmake_cfg.define("CMAKE_C_COMPILER", clang);
+        cmake_cfg.define("CMAKE_CXX_COMPILER", clangxx);
+        cmake_cfg.define("CMAKE_ASM_COMPILER", clang);
+        cmake_cfg.define("FIPS", "1");
+    }
+
+    if let Some(symbol_prefix) = build_prefix {
+        cmake_cfg.define("BORINGSSL_PREFIX", symbol_prefix);
+        let pwd = env::current_dir().unwrap();
+        let include_path = pwd.join(AWS_LC_PATH).join("include");
+        cmake_cfg.define(
+            "BORINGSSL_PREFIX_HEADERS",
+            include_path.display().to_string(),
+        );
+    }
+
+    cmake_cfg
+}
+
+fn build_aws_lc() -> Result<PathBuf, String> {
+    let mut cmake_cfg = prepare_cmake_build(false, Some(&prefix_string()));
+    let output_dir = cmake_cfg.build_target("ssl").build();
+
+    Ok(output_dir)
+}
```

Comment:
Why is this using a result? This never returns an error

Reply:
A previous iteration of this function had potential for errors. I will change the function signature.